tari-project/tari
# Implement transaction negotiation message resending in the wallet

## Description
Currently the logic for sending transactions and negotiating them between parties is not very robust. Each stage in the negotiation is sent only once. When sending we first try to send the message directly and if that fails, or discovery needs to occur, the message is sent indirectly via Store and Forward (SAF). We hoped to get SAF 100% reliable but due to reasons laid out in https://github.com/tari-project/tari/issues/2137 it is not tenable. 

This PR adds in the logic to periodically resends the initial Transaction Sender Message and the Transaction Reply in order to illicit the next step in the negotiation protocol from the counterparty. The PR also adds the logic to respond to repeated messages but with a cool down period in which repeats will be ignored to protect against DoS attacks. 

The PR also adds in a new comms level message to inform the counterparty if a transaction protocol has been cancelled. Initially this is sent if the Sender cancels an in progress transaction so that the receiver knows to stop expecting it. This message is also sent if a Transaction Reply is received for a cancelled message to let the Received know about the cancellation.

In order to stop a wallet resending these message indefinitely a transaction will be cancelled if it hasn’t resolved after a long timeout period (Default is being set to 3 days).

A large portion of the PR is tests and some code reorganization so its not as big as it looks.

## Motivation and Context
Closes https://github.com/tari-project/tari/issues/2137
Closes https://github.com/tari-project/tari/issues/1804

## How Has This Been Tested?
Tests are provided that test:
1. A sender repeats sending the initial message if no response is received from the receiver
2. A receiver repeats sending the Reply in no response is received from the receiver
3. For both parties that the repeated message is sent on startup if the period has elapsed.
4. For both parties that they can handle receiving the repeated messages and respond appropriately
5. Check that both parties ignore repeated messages sent before the cool down has elapsed
6. Check the cancellation messages are sent by the Sender when a transaction is cancelled
7. Check that receiver acts correctly on receiving a cancellation message
8. Check that the transaction is cancelled by both parties after the timeout is reached during operation or on startup.

## Types of changes

* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)

## Checklist:

* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [ ] I have squashed my commits into a single commit.
